test/EndToEnd/Tests/Infrastructure/LocalTestEnvironment.cs

```diff
@@ -0,0 +1,167 @@
+
+            // Stop the Dapr process
+            if (daprPids.TryGetValue(appName, out int daprPid))
+            {
+                this.logger.LogInformation($"Stopping Dapr sidecar for test app {appName}.");
+                Process.GetProcessById(daprPid).Kill();
+            }
+        }
+
+        /// <summary>
+        /// Starts the application in a container.
+        /// </summary>
+        /// <param name="appName">The name of the application.</param>
+        /// <param name="appPortOnHost">The port where the application should listen on the host.</param>
+        /// <param name="daprPortOnHost">The port where the Dapr sidecar should listen on the host.</param>
+        private async Task<string> StartAppContainerAsync(string appName, int appPortOnHost, int daprPortOnHost)
+        {
+            var imageName = $"{containerRegistry}/{appName}:{containerTag}";
+
+            // Pull the image
+            await dockerClient.Images.CreateImageAsync(
+                new DockerModels.ImagesCreateParameters
+                {
+                    FromImage = imageName
+                },
+                null,
+                new Progress<DockerModels.JSONMessage>((m) => logger.LogInformation($"{m.Status}: {m.Progress}, {m.ProgressMessage}"))
+            );
+
+            var containerParams = new DockerModels.CreateContainerParameters
+            {
+                Image = imageName,
+                Name = appName,
+                ExposedPorts = new Dictionary<string, DockerModels.EmptyStruct>
+                {
+                    { $"{appPortOnHost}", new DockerModels.EmptyStruct() }
+                },
+                HostConfig = new DockerModels.HostConfig
+                {
+                    PortBindings = new Dictionary<string, IList<DockerModels.PortBinding>>
+                    {
+                        { $"{appPortOnHost}", new List<DockerModels.PortBinding> { new DockerModels.PortBinding { HostPort = $"{appPortOnHost}" } } }
+                    }
+                },
+                Env = new List<string>
+                {
+                    $"DAPR_HTTP_HOST=host.docker.internal",
+                    $"DAPR_HTTP_PORT={daprPortOnHost}",
+                    $"ASPNETCORE_URLS=http://*:{appPortOnHost}",
+                }
+            };
+
+            // Create the container
+            var response = await dockerClient.Containers.CreateContainerAsync(containerParams);
+
+            if (response.Warnings != null)
+            {
+                foreach (var warning in response.Warnings)
+                {
+                    logger.LogWarning(warning);
+                }
+            }
+
+            // Start the container
+            var started = await dockerClient.Containers.StartContainerAsync(response.ID, new DockerModels.ContainerStartParameters());
+            if (!started)
+            {
+                throw new SystemException($"Failed to start container {response.ID}");
```

Comment:
Can we log the app name, port and host info as well. With multiple apps, it will help in figuring out which has failed.

Reply:
Great idea, done.